File: examples/play_file.pl
#!/usr/bin/env perl

# Jackalope Audio Engine
# Copyright 2019 Tyler Riddle <kg7oem@gmail.com>

#This source code is licensed according to the Perl Artistic License 2.0 of
#which you can find a copy in the doc/ subdirectory of this project.

#This license establishes the terms under which a given free software Package
#may be copied, modified, distributed, and/or redistributed.The intent is that
#the Copyright Holder maintains some artistic control over the development of
#that Package while still keeping the Package available as open source and
#free software.

#You are always permitted to make arrangements wholly outside of this license
#directly with the Copyright Holder of a given Package. If the terms of this
#license do not permit the full use that you propose to make of the Package,
#you should contact the Copyright Holder and seek a different licensing
#arrangement.

use strict;
use warnings;
use v5.10;

use Jackalope;

use constant SAMPLE_RATE => 48000;
use constant BUFFER_SIZE => 512;
use constant LADSPA_ZAMTUBE_ID => 1515476290;

main();

sub main {
    my @files = @ARGV;
    die "must specify one or more files to play" unless @files;

    Jackalope::init;

    foreach my $file (@files) {
        play_file($file);
    }
}

sub play_file {
    my ($filename) = @_;

    die "not a file: $filename" unless -f $filename;

    my $graph = Jackalope::Graph->new(
        "pcm.sample_rate" => SAMPLE_RATE,
        "pcm.buffer_size" => BUFFER_SIZE,
    );

    my $input_file = $graph->add_node(
        "object.type" => "audio::sndfile",
        "node.name" => "input file",
        "config.path", $filename,
    );

    my $system_audio = $graph->add_node(
        "object.type" => "audio::portaudio",
        "node.name" => "system audio",
        "sink.left" => "audio",
        "sink.right" => "audio",
    );

    my $left_tube = $graph->add_node(
        "object.type" => "audio::ladspa",
        "node.name" => "left tube",
        "plugin.id" => LADSPA_ZAMTUBE_ID,
    );

    my $right_tube = $graph->add_node(
        "object.type" => "audio::ladspa",
        "node.name" => "right tube",
        "plugin.id" => LADSPA_ZAMTUBE_ID,
    );

    $input_file->subscribe("object.stopped", $graph, "object.stop");

    if ($input_file->get_num_sources == 1) {
        $input_file->link("Output 1", $left_tube, "Audio Input 1");
        $input_file->link("Output 1", $right_tube, "Audio Input 1");
    } else {
        $input_file->link("Output 1", $left_tube, "Audio Input 1");
        $input_file->link("Output 2", $right_tube, "Audio Input 1");
    }

    $left_tube->link("Audio Output 1", $system_audio, "left");
    $right_tube->link("Audio Output 1", $system_audio, "right");

    $graph->run;

    return;
}

File: examples/play_file.cxx
// Jackalope Audio Engine
// Copyright 2019 Tyler Riddle <kg7oem@gmail.com>

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.

#include <chrono>
#include <iostream>
#include <string>

#include <jackalope/jackalope.h>
#include <jackalope/log/dest.h>
#include <jackalope/logging.h>
#include <jackalope/object.h>

#define BUFFER_SIZE 512
#define SAMPLE_RATE 48000
#define LADSPA_ZAMTUBE_ID 1515476290

using namespace jackalope::log;

jackalope_network_t make_tube_simulator(jackalope_graph_t& graph_in);

int main(int argc_in, char ** argv_in)
{
    if (argc_in != 2) {
        jackalope_panic("must specify exactly one audio file to play");
    }

    auto dest = jackalope::make_shared<console_dest_t>(level_t::info);
    get_engine()->add_destination(dest);

    jackalope_init();

    auto graph = jackalope_graph_t::make({
        { "pcm.sample_rate", jackalope::to_string(SAMPLE_RATE) },
        { "pcm.buffer_size", jackalope::to_string(BUFFER_SIZE) },
    });

    auto input_file = graph.make_node({
        { "object.type", "audio::sndfile" },
        { "node.name", "input file" },
        { "config.path", argv_in[1] },
    });

    auto system_audio = graph.make_node({
        { "object.type", "audio::portaudio" },
        { "node.name", "system audio" },
        { "sink.left", "audio" },
        { "sink.right", "audio" },
    });

    auto tube_simulator = make_tube_simulator(graph);

    input_file.subscribe(JACKALOPE_SIGNAL_OBJECT_STOPPED, graph, JACKALOPE_SLOT_OBJECT_STOP);

    input_file.link("Output 1", tube_simulator, "left");
    input_file.link("Output 1", tube_simulator, "right");

    tube_simulator.link("left", system_audio, "left");
    tube_simulator.link("right", system_audio, "right");

    tube_simulator.poke("config.drive", 0);
    tube_simulator.poke("config.gain", -3);

    graph.run();

    return(0);
}

jackalope_network_t make_tube_simulator(jackalope_graph_t& graph_in)
{
    auto tube_simulator = graph_in.make_network({
        { "node.name", "tube simulator" },
    });

    std::map<jackalope::string_t, jackalope::string_t> tube_property_name_map({
        { "config.Tube Drive", "config.drive" },
        { "config.Bass", "config.bass" },
        { "config.Mids", "config.mids" },
        { "config.Treble", "config.treble" },
    });

    for(auto property_name : { "config.gain", "config.drive", "config.bass", "config.mids", "config.treble" }) {
        tube_simulator.add_property(property_name, jackalope::property_t::type_t::real);
    }

    for(auto i : { "left", "right" }) {
        auto tube_node = tube_simulator.make_node({
            { "object.type", "audio::ladspa" },
            { "node.activate", "false" },
            { "node.name", jackalope::to_string(i, " tube") },
            { "plugin.id", jackalope::to_string(LADSPA_ZAMTUBE_ID) },
        });

        auto gain_node = tube_simulator.make_node({
            { "object.type", "audio::gain" },
            { "node.activate", "false" },
            { "node.name", jackalope::to_string(i, " gain") },
            { "pcm.sample_rate", jackalope::to_string(SAMPLE_RATE) },
            { "pcm.buffer_size", jackalope::to_string(BUFFER_SIZE) },
        });

        tube_simulator.add_source(i, "audio");
        tube_simulator.add_sink(i, "audio");

        for(auto i : tube_property_name_map) {
            tube_node.alias_property(i.first, tube_simulator, i.second);
        }

        gain_node.alias_property("config.gain", tube_simulator, "config.gain");

        tube_node.activate();
        gain_node.activate();

        tube_simulator.forward(i, tube_node, "Audio Input 1");
        tube_node.link("Audio Output 1", gain_node, "input");
        gain_node.forward("output", tube_simulator, i);
    }

    return tube_simulator;
}
